batching_kafka_consumer/__init__.py

```diff
@@ -129,6 +131,17 @@ def __init__(self, topics, worker, max_batch_size, max_batch_time, bootstrap_ser
         self.commit_log_topic = commit_log_topic
         self.dead_letter_topic = dead_letter_topic
 
+    def __record_timing(self, metric, value, tags=None, sample_rate=None):
+        if self.__metrics is None:
+            return
+
+        return self.__metrics.timing(
+            metric,
+            value,
+            tags=tags,
+            sample_rate=sample_rate if sample_rate is not None else self.__metrics_sample_rates.get(metric, 1),
+        )
```

Comment:
OK, so I don't love this solution, but I think it moves in the right direction.

The good: it consolidates all of the `if self.metrics is None` checks, we don't have to put every metric sample rate in the constructor and remember to keep it in sync any time the metrics change.

The bad: it only works with timings, and is easy to circumvent (accidentally, if you're not familiar with the codebase; intentionally, if for some reason you think that's a good idea.)

Ultimately, what I think would make sense is to provide a metrics wrapper on top of the dogstatsd API that has a the ability to provide per-metric sample rates as part of the constructor. Unfortunately this is going to have to be shared between here and Snuba at minimum (it'd also be nice to have it in Sentry and CDC, in my opinion.)

I also think that we should by default provide a no-op metrics backend that is API compatible with the dogstatsd API that just does nothing, so we don't have to litter the codebase with `if self.__metrics is None` checks.

Reply:
"Unfortunately this is going to have to be shared between here and Snuba at minimum (it'd also be nice to have it in Sentry and CDC, in my opinion.)"
There is a rudimentary one in CDC already but it is CDC specific. In order to reuse it it requires refactoring

Reply:
Also, are you sure you want to allow the call site to just override the rate ?
I cannot think about a case this should be expected considering this is a relatively simple class and it gets the sample rates when initialized (so we do not have to handle a config schema so far)

Reply:
> There is a rudimentary one in CDC already but it is CDC specific. In order to reuse it it requires refactoring

Yeah — I'm mostly just not sure where we'd put it, besides creating a new library strictly for the metrics wrapper and including it as a dependency in these repositories. That seems a little like overkill, but I guess there's not really a common thread otherwise.

> Also, are you sure you want to allow the call site to just override the rate ?

Good point — no, not really. I'll remove the option. It'd be confusing if it was used with different sample rates at different call sites anyway, since that'd effectively be weighting the sample anyway.

Reply:
Removed callsite sampling with 4a51ba9948cc29bb793cd26b9c6af36146f06545.